open-metadata-implementation/adapters/open-connectors/governance-daemon-connectors/open-lineage-connectors/open-lineage-janus-connector/src/main/java/org/odpi/openmetadata/openconnectors/governancedaemonconnectors/openlineageconnectors/janusconnector/buffergraph/MainGraphMapper.java

```diff
@@ -4,72 +4,302 @@
-        Iterator<Vertex> columnOutVertex = main.V().has(PROPERTY_KEY_ENTITY_GUID, columnOutGuid);
-        Iterator<Vertex> processVertex = main.V().has("id",processGuid);
+        if(mainG.V(columnInVertex.id()).outE(EDGE_LABEL_COLUMN_AND_PROCESS).inV().has(PROPERTY_KEY_ENTITY_GUID,processGuid).hasNext()){
+            return;
+        }
 
-        if (columnInVertex.hasNext() && columnOutVertex.hasNext()) {
+        if(columnInVertex != null && columnOutVertex != null) {
+            Vertex subProcess = mainG.addV(NODE_LABEL_SUB_PROCESS)
+                    .property(PROPERTY_KEY_ENTITY_NODE_ID, UUID.randomUUID().toString())
+                    .property(PROPERTY_KEY_ENTITY_GUID, processGuid)
+                    .property(PROPERTY_KEY_DISPLAY_NAME, processName)
+                    .next();
 
-            Vertex vertex = main.addV("SubProcess").next();
-            vertex.property("id", UUID.randomUUID());
-            vertex.property(PROPERTY_KEY_ENTITY_GUID, processGuid);
-            vertex.property(PROPERTY_KEY_ENTITY_NAME, processName);
+            columnInVertex.addEdge(EDGE_LABEL_COLUMN_AND_PROCESS, subProcess);
+            subProcess.addEdge(EDGE_LABEL_COLUMN_AND_PROCESS, columnOutVertex);
 
-            Vertex columnIn = columnInVertex.next();
-            Vertex columnOut = columnOutVertex.next();
+            Iterator<Vertex> processTopLevel = mainG.V().has(PROPERTY_KEY_ENTITY_NODE_ID,process.property(PROPERTY_KEY_ENTITY_GUID).value());
+            if(processTopLevel.hasNext()){
+                Vertex mainProcess  = processTopLevel.next();
+                subProcess.addEdge(EDGE_LABEL_SUBPROCESS_TO_PROCESS,mainProcess);
+                mainG.tx().commit();
 
-            columnIn.addEdge(NODE_LABEL_PROCESS, vertex);
-            vertex.addEdge(NODE_LABEL_PROCESS,columnOut);
+                addTableNode(columnInVertex,columnOutVertex,mainProcess);
 
-            if(processVertex.hasNext()){
-                Vertex processTopLevel = processVertex.next();
-                vertex.addEdge(NODE_LABEL_PROCESS,processTopLevel);
             }
             else {
-                Vertex mainProcess = main.addV("Process").next();
-                mainProcess.property("id", processGuid);
-                mainProcess.property(PROPERTY_KEY_ENTITY_NAME, processName);
+                Vertex mainProcess = mainG.addV(NODE_LABEL_PROCESS).next();
+                mainProcess.property(PROPERTY_KEY_ENTITY_NODE_ID, processGuid);
+                mainProcess.property(PROPERTY_KEY_ENTITY_GUID, processGuid);
+                mainProcess.property(PROPERTY_KEY_DISPLAY_NAME, processName);
+                subProcess.addEdge(EDGE_LABEL_SUBPROCESS_TO_PROCESS,mainProcess);
+
+                mainG.tx().commit();
 
-                vertex.addEdge(NODE_LABEL_PROCESS,mainProcess);
+                addTableNode(columnInVertex,columnOutVertex,mainProcess);
             }
+        }else{
+            log.debug("Process failed");
```

Comment:
Maybe should be log.error, and could use further information saying exactly which process failed